python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,652 @@
+
+    @property
+    def dim(self):
+        return self.index[0].dim()
+
+    @property
+    def size(self):
+        return self.index[0].size()
+
+    @property
+    def pq_dim(self):
+        return self.index[0].pq_dim()
+
+    @property
+    def pq_len(self):
+        return self.index[0].pq_len()
+
+    @property
+    def pq_bits(self):
+        return self.index[0].pq_bits()
+
+    @property
+    def metric(self):
+        return self.index[0].metric()
+
+    @property
+    def n_lists(self):
+        return self.index[0].n_lists()
+
+    @property
+    def rot_dim(self):
+        return self.index[0].rot_dim()
+
+    @property 
+    def codebook_kind(self):
+        return self.index[0].codebook_kind()
+
+
+
+@auto_sync_handle
+def build(IndexParams index_params, dataset, handle=None):
+    """
+    Builds an IVF-PQ index that can be later used for nearest neighbor search.
+
+    Parameters
+    ----------
+    index_params : IndexParams object
+    dataset : CUDA array interface compliant matrix shape (n_samples, dim)
+        Supported dtype [float, int8, uint8] 
+    {handle_docstring}
+
+    Returns
+    -------
+    index: ivf_pq.Index
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import cupy as cp
+
+        from pylibraft.common import Handle
+        from pylibraft.neighbors import ivf_pq
+
+        n_samples = 50000
+        n_features = 50
+        n_queries = 1000
+
+        dataset = cp.random.random_sample((n_samples, n_features), dtype=cp.float32)
+        handle = Handle()
+        index_params = ivf_pq.IndexParams(
+            n_lists=1024, 
+            metric="l2_expanded", 
+            pq_dim=10)
+        index = ivf_pq.build(index_params, dataset, handle=handle)
+
+        # Search using the built index
+        queries = cp.random.random_sample((n_queries, n_features), dtype=cp.float32)
+        k = 10
+        neighbors = cp.empty((n_queries, k), dtype=cp.uint64)
+        distances = cp.empty((n_queries, k), dtype=cp.float32)
+        ivf_pq.search(ivf_pq.SearchParams(), index, queries, k, neighbors, distances, handle=handle)
+
+        # pylibraft functions are often asynchronous so the
+        # handle needs to be explicitly synchronized
+        handle.sync()
+
+    """
+    dataset_cai = dataset.__cuda_array_interface__
+    dataset_dt = np.dtype(dataset_cai["typestr"])
+    _check_input_array(dataset_cai,  [np.dtype('float32'), np.dtype('byte'), np.dtype('ubyte')])
+    cdef uintptr_t dataset_ptr = dataset_cai["data"][0]
+        
+    cdef uint64_t n_rows = dataset_cai["shape"][0]
+    cdef uint32_t dim = dataset_cai["shape"][1]
+    
+    if handle is None:
+        handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+        
+    idx = Index()
+        
+    if dataset_dt == np.float32:
+        c_ivf_pq.build(deref(handle_),      
+                       index_params.params,     
```

Comment:
Do you think it would make sense to wrap this into interruptible (https://github.com/rapidsai/raft/blob/branch-22.12/python/pylibraft/pylibraft/common/interruptible.pyx#L36-L54) to allow ctrl+c the build/search? Also it may be useful to set release the GIL for all member of the index as in https://github.com/rapidsai/cuml/blob/branch-22.12/python/cuml/svm/linear.pyx#L47

Reply:
@achirkin the `handle.sync()` in the Python API invokes `handle.sync_stream()` in the C++ layer. Isn't that enough to wrap this in the interruptible call? Or do we need to add `nogil` to all of the cython externs in order for it to work?

Reply:
I think, it's not enough, but also we'd need to do the same `interruptible` wrapper for the `handle.sync_stream()` to be able to reliably cancel the task.

Here's my intuition:
-  The cpu thread gets blocked by any synchronization call. To make it interruptible, we need this point to `interruptible::synchronize` (which is the case for `handle.sync_stream()`).
- To make a program to respond to ctrl+c, we need to install an interrupt handler; this is what [`with cuda_interruptible():`](https://github.com/rapidsai/raft/blob/branch-22.12/python/pylibraft/pylibraft/common/interruptible.pyx#L36-L54) snipped does. At the moment, raft's python wrapper on the `handle.sync_stream()` does not have it (but it should!).
- If the C++-side function does not contain any synchronization points, there's nothing to interrupt. If it was the case for ivf-pq functions, calling the `handle.sync_stream()` (patched with the interrupt handler) would indeed be enough for the interruptible behavior, because the program would exit the ifv-pq functions and wait on the latter. At the moment ivf-pq has one synchronization point in `search` (very fast, shouldn't be a problem) and plenty of them in `build` (this is what we really want to be able to cancel on ctrl+c).
- Honestly, I'm not entirely sure if releasing the GIL for any of these C++ calls is necessary. Yet, even if it's not the case, I think there may be some benefits in setting everything we can to `nogil` (perhaps, for running concurrent computations?..)